File: leetcode/dp/2403_Minimum_Time_to_Kill_All_Monsters/codes/dp.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int n;
    vector<int> p;
    vector<long long> memo;

    long long dfs(int mask) {
        if (mask == (1 << n) - 1) return 0;
        long long &res = memo[mask];
        if (res != -1) return res;
        int g = 1 + __builtin_popcount((unsigned)mask);
        long long best = LLONG_MAX;
        for (int i = 0; i < n; i++) {
            if (!((mask >> i) & 1)) {
                long long days = (p[i] + (long long)g - 1) / g;
                long long v = days + dfs(mask | (1 << i));
                if (v < best) best = v;
            }
        }
        return res = best;
    }

    long long minimumTime(vector<int>& power) {
        p = power;
        n = (int)p.size();
        memo.assign(1 << n, -1);
        return dfs(0);
    }
};

File: leetcode/dp/2403_Minimum_Time_to_Kill_All_Monsters/codes/dp.py
from typing import List
from functools import lru_cache
import math

class Solution:
    def minimumTime(self, power: List[int]) -> int:
        n = len(power)

        @lru_cache(None)
        def dp(mask: int, gain: int) -> int:
            if mask == (1 << n) - 1:
                return 0
            ans = float('inf')
            for i in range(n):
                if not (mask & (1 << i)):
                    days = math.ceil(power[i] / gain)
                    ans = min(ans, days + dp(mask | (1 << i), gain + 1))
            return ans

        return dp(0, 1)
